### PR TITLE
Online DDL: configurable cut-over threshold

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

In https://github.com/vitessio/vitess/pull/12451 we increased `vreplicationCutOverThreshold` from `5sec` to `10sec`. We now make this variable configurable via command line flag.

We also rename it as `cutOverThreshold` ; which basically stands for a few timeout-related operations during the vreplication cut-over phase and most importantly to the cut-over operation itself.

This PR also takes care of query buffering timeout. As reminder, query buffering was introduced in https://github.com/vitessio/vitess/pull/9755, specifically to buffer queries during an Online DDL cut-over. That buffering also has a timeout: `maxQueryBufferDuration`.

But here's a problem: `maxQueryBufferDuration` has to be greater than, or equal, to  `vreplicationCutOverThreshold`, because we want to buffer queries for the duration of the cut-over.  And so in https://github.com/vitessio/vitess/pull/12451, when we increased `vreplicationCutOverThreshold` to `10sec`, we also had to manually increase `maxQueryBufferDuration` from `10sec` to `15sec`. That's an unhealthy dependency to manage.

as of this PR, the buffering timeout is controlled by the process who places the buffering rule. So when you create a rule, you provide a `cancelCtx`, as before, that means "I'm done with this buffering rule, no need to further buffer anything", and now also a `timeout`, that means "if the query is buffered longer than this, fail the query".

So now the Online DDL imposes a buffering rule with:

- A `cutOverThreshold` that is configurable via `--cut-over-threshold=...` strategy flag
- A timeout based on `cutOverThreshold` + `5sec` as nice margins,

Example: 
```
vtctldclient  ApplySchema --ddl-strategy "vitess --cut-over-threshold=20s" --sql "alter table corder engine=innodb" commerce
```

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Legit endtoend test error:
```
2023-03-09T14:06:24.7159723Z         	Error:      	Expected nil, but got: &mysql.SQLError{Num:0x451, State:"HY000", Message:"target: ks.1.primary: vttablet: rpc error: code = FailedPrecondition desc = buffer timeout in rule: buffered for cut-over (CallerID: userData1)", Query:"\n\t\tINSERT IGNORE INTO stress_test (id, rand_val) VALUES (2180, left(md5(rand()), 8))\n\t"}
```
caused by changes in this PR

**Comment by shlomi-noach** [comment]
Fixed issue leading to endtoend failure

**Comment by shlomi-noach** [comment]
Ready for review

**Comment by shlomi-noach** [comment]
I confess I'm not sure whether new variables should be introduced with dashes or underscores.

**Comment by shlomi-noach** [comment]
Documentation PR: https://github.com/vitessio/website/pull/1407

**Comment by shlomi-noach** [comment]
Ping for review, this has been out for 3 weeks now :pray:

File: go/vt/vttablet/tabletserver/rules/rules.go [hunk 1]
```diff
@@ -329,6 +339,9 @@ func (qr *Rule) MarshalJSON() ([]byte, error) {
 	if qr.act != QRContinue {
 		safeEncode(b, `,"Action":`, qr.act)
 	}
+	if qr.timeout != 0 {
+		safeEncode(b, `,"Timeut":`, qr.timeout)
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletserver/rules/rules.go:343]: Typo: Timeut => Timeout

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 2]
```diff
@@ -562,7 +562,7 @@ func (qre *QueryExecutor) checkPermissions() error {
 				// good! We have buffered the query, and buffering is completed
 			case <-bufferingTimeoutCtx.Done():
 				// Sorry, timeout while waiting for buffering to complete
-				return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "buffer timeout in rule: %s", desc)
+				return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "buffer timeout after in rule after %s: %s", timeout.String(), desc)
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletserver/query_executor.go:565]: did you want the two "after"s in the error message here?

**Review by rohit-nayak-ps — APPROVED** [review]
LGTM

A couple of minor nits, which you may want to fix now or in a subsequent pr ...

File: go/vt/vttablet/tabletserver/rules/rules.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/rules/rules.go]: fixed

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/query_executor.go]: fixed

File: go/flags/endtoend/vttablet.txt [hunk 3]
```diff
@@ -195,6 +195,7 @@ Usage of vttablet:
       --mysqlctl_mycnf_template string                                   template file to use for generating the my.cnf file during server init
       --mysqlctl_socket string                                           socket file to use for remote mysqlctl actions (empty for local actions)
       --onclose_timeout duration                                         wait no more than this for OnClose handlers before stopping (default 10s)
+      --online-ddl-cut-over-threshold duration                           max time to wait on migration cut-over before cancelling (default 10s)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/flags/endtoend/vttablet.txt:198]: I like the idea. Let me work on that. Parsing values in DDL strategy is weaker than using `flag`/`pflag` but let's see what can be done.

**Comment by shlomi-noach** [comment]
Refactored! Cut-over timeout is now configurable via `vitess --cut-over-timeout` DDL strategy ; so this is configurable per-migration. The `vttablet` flag is removed.

**Comment by shlomi-noach** [comment]
`Upgrade Downgrade` tests failing an an irrelevant issue: `golang` version

File: go/vt/vttablet/onlineddl/executor.go [hunk 4]
```diff
@@ -231,6 +230,15 @@ func newGCTableRetainTime() time.Time {
 	return time.Now().UTC().Add(retainOnlineDDLTables)
 }
 
+// getMigrationCutOverThreshold returns the cut-over threshold for the given migration. The migration's
+// DDL Strategy may excplicitly set the threshold; otherwise, we return the default cut-over threshold.
+func getMigrationCutOverThreshold(onlineDDL *schema.OnlineDDL) time.Duration {
+	if threshold, _ := onlineDDL.StrategySetting().CutOverThreshold(); threshold != 0 {
+		return threshold
+	}
+	return cutOverThreshold
```
Comment (mattlord) [inline go/vt/vttablet/onlineddl/executor.go:239]: I feel like this variable should be called defaultCutOverThreshold, but that's obviously very nitty.

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 5]
```diff
@@ -562,7 +562,7 @@ func (qre *QueryExecutor) checkPermissions() error {
 				// good! We have buffered the query, and buffering is completed
 			case <-bufferingTimeoutCtx.Done():
 				// Sorry, timeout while waiting for buffering to complete
-				return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "buffer timeout after in rule after %s: %s", timeout.String(), desc)
+				return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "buffer timeout after %v in rule: %s", timeout.String(), desc)
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/query_executor.go:565]: Nit, but I think you can use %v without .String() or %s with .String.

**Review by mattlord — APPROVED** [review]
Thanks! ❤️

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/query_executor.go]: Fixed

File: go/vt/vttablet/onlineddl/executor.go (hunk 4 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go]: Renamed

**Pull request merged**